cclib/method/ddec.py

```diff
@@ -300,6 +327,324 @@ def calculate_refcharges(self):
+            # psi_A^I -- Equation S100 in doi: 10.1039/c6ra04656h
+            psiAI[atomi] = (
+                self._integrate_from_radial([self._cond_density[atomi]], [atomi])
+                - self.data.atomnos[atomi]
+                + self.refcharges[-1][atomi]
+            )
+
+            self._candidates_bigPsi.append([bigpsiAI[atomi]])
+            self._candidates_psi.append([psiAI[atomi]])
+
+            fitpsi = True  # when convergence is reached, this is modified to False.
+
+            while psiAI[atomi] <= 0:
+                # Iterative algorithm until convergence
+                # Refer to S101 in doi: 10.1039/c6ra04656h
+                bigpsiAI[atomi] = 2 * bigpsiAI[atomi] - psiAI[atomi] / self._integrate_from_radial(
+                    [numpy.sqrt(self._y_a[atomi])], [atomi]
+                )
+
+                # When Psi is updated, related quantities are updated as well
+                # Refer to S100 in doi: 10.1039/c6ra04656h
+                psiAI[atomi], self._cond_density[atomi] = self._psiai(
+                    self._y_a[atomi], bigpsiAI[atomi], atomi
+                )
+
+                self._candidates_psi[atomi].append(psiAI[atomi])
+                self._candidates_bigPsi[atomi].append(bigpsiAI[atomi])
+
+                if abs(psiAI[atomi]) < 1e-10:
+                    fitpsi = False
+                    break
+
+            # lowerbigPsi is largest negative Psi.
+            # upperbigPsi is smallest positive Psi.
+            if fitpsi:
+                self._candidates_psi[atomi] = numpy.array(self._candidates_psi[atomi], dtype=float)
+                self._candidates_bigPsi[atomi] = numpy.array(
+                    self._candidates_bigPsi[atomi], dtype=float
+                )
+                if numpy.count_nonzero(self._candidates_psi[atomi] < 0) > 0:
+                    lower_ind = numpy.where(
+                        self._candidates_psi[atomi]
+                        == self._candidates_psi[atomi][self._candidates_psi[atomi] < 0].max()
+                    )[0][0]
+                    lowerbigPsi = self._candidates_bigPsi[atomi][lower_ind]
+                    lowerpsi = self._candidates_psi[atomi][lower_ind]
+                else:  # assign some large negative number
+                    lowerbigPsi = -9999999.0
+                    lowerpsi = -9999999.0
+                if numpy.count_nonzero(self._candidates_psi[atomi] > 0) > 0:
+                    upper_ind = numpy.where(
+                        self._candidates_psi[atomi]
+                        == self._candidates_psi[atomi][self._candidates_psi[atomi] > 0].min()
+                    )[0][0]
+                    upperbigPsi = self._candidates_bigPsi[atomi][upper_ind]
+                    upperpsi = self._candidates_psi[atomi][upper_ind]
+                else:  # assign some large positive number
+                    upperbigPsi = 9999999.0
```

Comment:
can we use something like np.inf here and above

Reply:
I've used np.PINF and np.NINF which were suggested by Eric in a comment below this. Thanks!